adservertargeting/reqdataprocessor.go

```diff
@@ -0,0 +1,90 @@
+package adservertargeting
+
+import (
+	"encoding/json"
+	"github.com/buger/jsonparser"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+func getValueFromBidRequest(dataHolder *reqImpCache, path string, ampData map[string]string) (RequestTargetingData, error) {
+	//use the path specified in 'value' to look for data in the ortb bidrequest.
+	res := RequestTargetingData{}
+
+	//check if key points to query param from ext.prebid.amp.data
+	ampDataValue, err := getValueFromQueryParam(path, ampData)
+	if ampDataValue != nil || err != nil {
+		res.SingleVal = ampDataValue
+		return res, err
+	}
+
+	// check if key points to imp data
+	impData, err := getValueFromImp(path, dataHolder)
+	if len(impData) > 0 || err != nil {
+		res.ImpData = impData
+		return res, err
+	}
+
+	// get data by key from request
+	requestValue, err := getDataFromRequest(path, dataHolder)
+	if requestValue != nil || err != nil {
+		res.SingleVal = requestValue
+		return res, err
+	}
```

Comment:
Right, value can be found in one of: query params, imps or request. 
My idea here is to hide string splits inside of `getValueFromQueryParam`, `getValueFromImp`, and `getDataFromRequest`. If value is found in one of them - func getValueFromBidRequest doesn't look for value anywhere else and returns it.  
`getDataFromRequest` should be executed at the end, after everything else, because it's not clear how it should start. 

My initial version was parsing strings and it was overloaded by if cases. I don't think this approach will make it cleaner. 
Unless I miss something.